src/jit/optcse.cpp

```diff
@@ -880,39 +972,103 @@ void Compiler::optValnumCSE_InitDataFlow()
+        for (unsigned inx = 0; inx < m_comp->optCSECandidateCount; inx++)
+        {
+            unsigned CseAvailBit = inx * 2;
+
+            // a one preserves availability and a zero kills the availability
+            // we generate this kind of bit pattern:  101010101010
+            //
+            BitVecOps::AddElemD(m_comp->cseLivenessTraits, m_callKills, CseAvailBit);
+        }
     }
 
     // At the start of the merge function of the dataflow equations, initialize premerge state (to detect changes.)
     void StartMerge(BasicBlock* block)
     {
-        BitVecOps::Assign(m_pBitVecTraits, m_preMergeOut, block->bbCseOut);
+        BitVecOps::Assign(m_comp->cseLivenessTraits, m_preMergeOut, block->bbCseOut);
+#ifdef DEBUG
+        if (m_comp->verbose)
+        {
+            printf("StartMerge BB%02u\n", block->bbNum);
+            printf("  :: cseOut    = %s\n", genES2str(m_comp->cseLivenessTraits, block->bbCseOut));
+        }
+#endif // DEBUG
     }
 
     // During merge, perform the actual merging of the predecessor's (since this is a forward analysis) dataflow flags.
     void Merge(BasicBlock* block, BasicBlock* predBlock, flowList* preds)
     {
-        BitVecOps::IntersectionD(m_pBitVecTraits, block->bbCseIn, predBlock->bbCseOut);
+#ifdef DEBUG
+        if (m_comp->verbose)
+        {
+            printf("Merge BB%02u and BB%02u\n", block->bbNum, predBlock->bbNum);
+            printf("  :: cseIn     = %s\n", genES2str(m_comp->cseLivenessTraits, block->bbCseIn));
+            printf("  :: cseOut    = %s\n", genES2str(m_comp->cseLivenessTraits, block->bbCseOut));
+        }
+#endif // DEBUG
+
+        BitVecOps::IntersectionD(m_comp->cseLivenessTraits, block->bbCseIn, predBlock->bbCseOut);
+
+#ifdef DEBUG
+        if (m_comp->verbose)
+        {
+            printf("  => cseIn     = %s\n", genES2str(m_comp->cseLivenessTraits, block->bbCseIn));
+        }
+#endif // DEBUG
     }
 
     // At the end of the merge store results of the dataflow equations, in a postmerge state.
     bool EndMerge(BasicBlock* block)
     {
-        BitVecOps::DataFlowD(m_pBitVecTraits, block->bbCseOut, block->bbCseGen, block->bbCseIn);
-        return !BitVecOps::Equal(m_pBitVecTraits, block->bbCseOut, m_preMergeOut);
+        EXPSET_TP cseIn_withCallsKill = BitVecOps::MakeEmpty(m_comp->cseLivenessTraits);
+
+        if (!BitVecOps::IsEmpty(m_comp->cseLivenessTraits, block->bbCseIn))
+        {
+            BitVecOps::Assign(m_comp->cseLivenessTraits, cseIn_withCallsKill, block->bbCseIn);
+            if (block->bbFlags & BBF_HAS_CALL)
+            {
+                BitVecOps::IntersectionD(m_comp->cseLivenessTraits, cseIn_withCallsKill, m_callKills);
+            }
+        }
+        BitVecOps::DataFlowD(m_comp->cseLivenessTraits, block->bbCseOut, block->bbCseGen, cseIn_withCallsKill);
+
+        bool result = !BitVecOps::Equal(m_comp->cseLivenessTraits, block->bbCseOut, m_preMergeOut);
```

Comment:
This is why we need the extra bit:

This dataflow algorithm will push all of this block successors onto the `worklist `to visit when this `EndMerge` step returns true:

In dataflow.h line 63:
```
        if (callback.EndMerge(block))
        {
            for (BasicBlock* succ : block->GetAllSuccs(m_pCompiler))
            {
                worklist.insert(worklist.end(), succ);
            }
        }
```

If the very first time that we vist this block we compute that `block->bbCseOut`, has all bits set, (i.e `cseFull`)
Then we won't visit this blocks successors at all.

Before my changes here, we always used `unsigned (-1) ` as the value for cseFull, this meant that we essentially always had one or more extra bits set in `bbCseOut`.  

I found this issue after my change because there were a couple of methods that had blocks that made every CSE available and then the algorithm didn't properly visit successor blocks.
`